### ISSUE TITLE
[Bug] Assertion error "lora_act_out.shape[1] == rank_down" related to LoRAs with missing blocks

### ISSUE THREAD

**Issue by EricBCoding** [mit-han-lab/nunchaku#278]
### Checklist

- [x] 1. I have searched for related issues and FAQs (https://github.com/mit-han-lab/nunchaku/discussions/262) but was unable to find a solution.
- [x] 2. The issue persists in the latest version.
- [x] 3. Please note that without environment information and a minimal reproducible example, it will be difficult for us to reproduce and address the issue, which may delay our response.
- [x] 4. If your report is a question rather than a bug, please submit it as a discussion at https://github.com/mit-han-lab/nunchaku/discussions/new/choose. Otherwise, this issue will be closed.
- [x] 5. If this is related to ComfyUI, please report it at https://github.com/mit-han-lab/ComfyUI-nunchaku/issues.
- [x] 6. I will do my best to describe the issue in English.

### Describe the Bug

Hi,

I see that #265 is already on your roadmap, but I'm running into a different error when trying to load LoRAs that have been trained on specific blocks [as popularized here](https://old.reddit.com/r/StableDiffusion/comments/1f523bd/good_flux_loras_can_be_less_than_45mb_128_dim/). Essentially, the LoRA only contains information for single blocks 7 and 20 and has an extremely high rank and alpha of 1536.

The error is as follows:

```
2025-04-14 17:18:01.985] [info] Loading partial weights from pytorch
[2025-04-14 17:18:02.528] [info] Done.
Assertion failed: lora_act_out.shape[1] == rank_down, file C:\Users\muyang\Desktop\nunchaku-dev\src\kernels\zgemm\gemm_w4a4_launch_impl.cuh, line 242
```

I have attempted to pad the LoRA with the missing keys, causing it to grow to 14 GB. I'm no longer seeing an assertion error, but my computer has slowed to a crawl. (0/20 inference steps completed in the last 5 minutes.)

### Environment

Python 3.10, PyTorch 2.5, CUDA 12.4

### Reproduction Steps

N/A

**Comment by liaozeyu20000110** [comment]
A similar error message！

What am I doing：
1. I tried to override every inference function in the source code, then recompile. 
2. And load custom lora (it's from  [OminiControl](https://github.com/Yuanshi9815/OminiControl) )
3. I found that I can load the official sample of aleksa-codes/flux-ghibsky-illustration and get the results, However, the following error message is displayed when the custom lora is loaded
4. According to my comparison, the difference between the two LORAs is the imposition of lora control on different modules and layers, is that the reason？
 
Error message：
` python: ./nunchaku/src/kernels/zgemm/gemm_w4a4_launch_impl.cuh:242: nunchaku::kernels::GEMM_W4A4_Launch<nunchaku::kernels::GEMMConfig_W4A4<true>, false>::gemm_w4a4(Tensor, Tensor, Tensor, Tensor, Tensor, Tensor, Tensor, Tensor, Tensor, Tensor, Tensor, Tensor, Tensor, Tensor, Tensor, Tensor, Tensor, Tensor, Tensor, bool, std::vector<float>, bool, bool, float, Tensor, Tensor, Tensor, Tensor, int)::<lambda(typename NextEpilogue::Arguments, typename MidEpilogue::Arguments)> [with NextEpilogue = nunchaku::kernels::GEMM_W4A4<nunchaku::kernels::GEMMConfig_W4A4<true> >::EpilogueQuantize<false, true, false>; MidEpilogue = nunchaku::kernels::Epilogues<nunchaku::kernels::GEMMConfig_W4A4<true> >::EpilogueGelu; typename NextEpilogue::Arguments = nunchaku::kernels::GEMM_W4A4<nunchaku::kernels::GEMMConfig_W4A4<true> >::EpilogueQuantize<false, true, false>::Arguments; typename MidEpilogue::Arguments = nunchaku::kernels::Epilogues<nunchaku::kernels::GEMMConfig_W4A4<true> >::EpilogueGelu::Arguments]: Assertion lora_act_out.shape[1] == rank_down failed.`

I would appreciate it if I could get some help ！ ^^

**Comment by lmxyy** [comment]
For debugging purposes, could you share your LoRAs along with usage instructions? Also, which tool or library did you use to train them?

**Comment by EricBCoding** [comment]
Hi @lmxyy,

I recently upgraded from CUDA 12.4 to 12.6 and am happy to report that this issue is no longer occurring for me. Perhaps a change in ComfyUI or some other dependency helped out.

---

I cannot distribute this particular LoRA, but I will share some information about my setup in case you want to investigate compatibility with CUDA 12.4.

The LoRA was trained with [kohya-ss/sd-scripts](https://github.com/kohya-ss/sd-scripts) using the following options:

```bat
set PROJECT="flux_project_example"
set REVISION="16"
set RANK=1536
call ./venv/Scripts/activate

accelerate launch flux_train_network.py ^
	--pretrained_model_name_or_path="T:/code/python/comfyui/models/unet/jibMixFlux_v8AccentueightNow.safetensors" ^
	--clip_l "T:/code/python/comfyui/models/clip/clip_l.safetensors" ^
	--t5xxl "T:/code/python/comfyui/models/clip/t5xxl_fp16.safetensors" ^
	--ae "T:/code/python/comfyui/models/vae/ae.sft" ^
	--dataset_config="./_flux_in/%PROJECT%/config.toml" ^
	--output_dir="./_flux_out/%PROJECT%_%REVISION%" ^
	--output_name="%PROJECT%_%REVISION%" ^
	--network_module=networks.lora_flux ^
	--resolution="1024,1024" ^
	--save_model_as="safetensors" ^
	--max_train_steps=4000 ^
	--save_every_n_steps=1000  ^
	--save_every_n_epochs=15 ^
	--network_dim=%RANK% ^
	--network_alpha=%RANK% ^
	--no_half_vae ^
	--gradient_checkpointing ^
	--persistent_data_loader_workers ^
	--enable_bucket ^
	--random_crop ^
	--bucket_reso_steps=64 ^
	--min_bucket_reso=512 ^
	--xformers ^
	--mixed_precision="bf16" ^
	--caption_extension=".txt" ^
	--optimizer_type="prodigy" ^
	--lr_scheduler="cosine" ^
	--lr_warmup_steps=150 ^
	--prior_loss_weight=1.0 ^
	--max_grad_norm=1.0 ^
	--learning_rate=1.0 ^
	--seed=0 ^
	--network_train_unet_only ^
	--mixed_precision bf16 ^
	--highvram ^
	--network_args "train_double_block_indices=none" "train_single_block_indices=7,20" "single_mod_dim=0" "single_dim=0" ^
	--xformers ^
	--fp8_base ^
	--full_bf16 ^
	--timestep_sampling="shift" ^
	--discrete_flow_shift=3.1582 ^
	--model_prediction_type="raw" ^
	--guidance_scale=1.0 ^
	--sdpa ^
	--apply_t5_attn_mask ^
	--optimizer_args weight_decay=1e-04 betas=(0.9,0.9999) eps=1e-08 decouple=True use_bias_correction=True safeguard_warmup=True beta3=None
```

If it would help, I can try training another LoRA with this setup. Is there an example dataset/keyword you'd like me to use?

**Comment by lmxyy** [comment]
> Hi [@lmxyy](https://github.com/lmxyy),
> 
> I recently upgraded from CUDA 12.4 to 12.6 and am happy to report that this issue is no longer occurring for me. Perhaps a change in ComfyUI or some other dependency helped out.
> 
> I cannot distribute this particular LoRA, but I will share some information about my setup in case you want to investigate compatibility with CUDA 12.4.
> 
> The LoRA was trained with [kohya-ss/sd-scripts](https://github.com/kohya-ss/sd-scripts) using the following options:
> 
> set PROJECT="flux_project_example"
> set REVISION="16"
> set RANK=1536
> call ./venv/Scripts/activate
> 
> accelerate launch flux_train_network.py ^
> 	--pretrained_model_name_or_path="T:/code/python/comfyui/models/unet/jibMixFlux_v8AccentueightNow.safetensors" ^
> 	--clip_l "T:/code/python/comfyui/models/clip/clip_l.safetensors" ^
> 	--t5xxl "T:/code/python/comfyui/models/clip/t5xxl_fp16.safetensors" ^
> 	--ae "T:/code/python/comfyui/models/vae/ae.sft" ^
> 	--dataset_config="./_flux_in/%PROJECT%/config.toml" ^
> 	--output_dir="./_flux_out/%PROJECT%_%REVISION%" ^
> 	--output_name="%PROJECT%_%REVISION%" ^
> 	--network_module=networks.lora_flux ^
> 	--resolution="1024,1024" ^
> 	--save_model_as="safetensors" ^
> 	--max_train_steps=4000 ^
> 	--save_every_n_steps=1000  ^
> 	--save_every_n_epochs=15 ^
> 	--network_dim=%RANK% ^
> 	--network_alpha=%RANK% ^
> 	--no_half_vae ^
> 	--gradient_checkpointing ^
> 	--persistent_data_loader_workers ^
> 	--enable_bucket ^
> 	--random_crop ^
> 	--bucket_reso_steps=64 ^
> 	--min_bucket_reso=512 ^
> 	--xformers ^
> 	--mixed_precision="bf16" ^
> 	--caption_extension=".txt" ^
> 	--optimizer_type="prodigy" ^
> 	--lr_scheduler="cosine" ^
> 	--lr_warmup_steps=150 ^
> 	--prior_loss_weight=1.0 ^
> 	--max_grad_norm=1.0 ^
> 	--learning_rate=1.0 ^
> 	--seed=0 ^
> 	--network_train_unet_only ^
> 	--mixed_precision bf16 ^
> 	--highvram ^
> 	--network_args "train_double_block_indices=none" "train_single_block_indices=7,20" "single_mod_dim=0" "single_dim=0" ^
> 	--xformers ^
> 	--fp8_base ^
> 	--full_bf16 ^
> 	--timestep_sampling="shift" ^
> 	--discrete_flow_shift=3.1582 ^
> 	--model_prediction_type="raw" ^
> 	--guidance_scale=1.0 ^
> 	--sdpa ^
> 	--apply_t5_attn_mask ^
> 	--optimizer_args weight_decay=1e-04 betas=(0.9,0.9999) eps=1e-08 decouple=True use_bias_correction=True safeguard_warmup=True beta3=None
> If it would help, I can try training another LoRA with this setup. Is there an example dataset/keyword you'd like me to use?

You can simply train a LoRA with which I can easily reproduce the error. Please share the prompt, any other hyperparameters you used, and the expected output image.

**Comment by lmxyy** [comment]
As there have been no further responses from the issuer, I will proceed to close this issue.
